pkg/gc/gc_state_manager.go

```diff
@@ -0,0 +1,567 @@
+//
+// Returns a struct AdvanceTxnSafePointResult, which contains the old txn safe point, the target, and the new
+// txn safe point it finally made it to advance to. If there's something blocking the txn safe point from being
+// advanced to the given target, it may finally be advanced to a smaller value or remains the previous value, in which
+// case the BlockerDescription field of the AdvanceTxnSafePointResult will be set to a non-empty string describing
+// the reason.
+//
+// Txn safe point of a single keyspace should never decrease. If the given target is smaller than the previous value,
+// it returns an error.
+//
+// WARNING: This method is only used to manage the GC procedure, and should never be called by code that doesn't
+// have the responsibility to manage GC. It can only be called on NullKeyspace or keyspaces with keyspace level GC
+// enabled.
+func (m *GCStateManager) AdvanceTxnSafePoint(keyspaceID uint32, target uint64, now time.Time) (AdvanceTxnSafePointResult, error) {
+	keyspaceID, err := m.redirectKeyspace(keyspaceID, false)
+	if err != nil {
+		return AdvanceTxnSafePointResult{}, err
+	}
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	return m.advanceTxnSafePointImpl(keyspaceID, target, now)
+}
+
+// advanceTxnSafePointImpl is the internal implementation of AdvanceTxnSafePoint, assuming keyspaceID has been checked
+// and the mutex has been acquired.
+func (m *GCStateManager) advanceTxnSafePointImpl(keyspaceID uint32, target uint64, now time.Time) (AdvanceTxnSafePointResult, error) {
+	// Marks whether it's needed to provide the compatibility for old versions.
+	//
+	// In old versions, every time TiDB performs GC, it updates the service safe point of "gc_worker" new txn safe
+	// point.
+	// Note that in old versions, there wasn't the concept of txn safe point. The step to update the service safe
+	// point of "gc_worker" is somewhat just like the current procedure of advancing the txn safe point, the most
+	// important purpose of which is to find the actual GC safe point that's safe to use.
+	downgradeCompatibleMode := false
+
+	var oldTxnSafePoint uint64
+	newTxnSafePoint := target
+	minBlocker := target
+	var blockingBarrier *endpoint.GCBarrier
+	var blockingMinStartTSOwner *string
+
+	err := m.gcMetaStorage.RunInGCStateTransaction(func(wb *endpoint.GCStateWriteBatch) error {
+		var err1 error
+		oldTxnSafePoint, err1 = m.gcMetaStorage.LoadTxnSafePoint(keyspaceID)
+		if err1 != nil {
+			return err1
+		}
+
+		if target < oldTxnSafePoint {
+			return errs.ErrDecreasingTxnSafePoint.GenWithStackByArgs(oldTxnSafePoint, target)
+		}
+
+		barriers, err1 := m.gcMetaStorage.LoadAllGCBarriers(keyspaceID)
+		if err1 != nil {
+			return err1
+		}
+
+		for _, barrier := range barriers {
+			if keyspaceID == constant.NullKeyspaceID && barrier.BarrierID == keypath.GCWorkerServiceSafePointID {
+				downgradeCompatibleMode = true
+				continue
+			}
+
+			if barrier.IsExpired(now) {
```

Comment:
I noticed that this comparison occurs during/after the etcd transcation. If certain disk failures cause the read/write operation to take an extremely long time, would this affect the judgment of the barrier time?

Reply:
I didn't get it. Can you explain the bad case you are concerning about in detail?
In my consideration, it's safe that the actual expiration time is a little longer than the specified expiration time, and guarantee a GC barrier to be expired once it's regarded expired (delete immediately in the currrent transaction if `IsExpired` returns true). It should not affect the correctness or safety when a transaction runs too long.